lib/services/resource_sharer.rb

```diff
@@ -0,0 +1,88 @@
+class ResourceSharer
+  WHITELISTED_RESOURCE_TYPES = %w(
+    MiqTemplate
+    ServiceTemplate
+  ).freeze
+
+  include ActiveModel::Model
+
+  attr_accessor :user, :resource, :tenants, :features
+
+  with_options :presence => true do
+    validates :user
+    validates :resource
+    validates :tenants
+    validates :features
+  end
+
+  validate :allowed_resource_type
+  validate :rbac_visibility
+  validate :valid_tenants
+  validate :allowed_features
+
+  ##
+  # Creates shares from the user with +features+ to +tenants+ with the given +resource+
+  # @param user - The user sharing the resource
+  # @param resource - The resource to be shared
+  # @param tenants - The tenants to share the resource with
+  # @param features - The product features to be associated with the share. Features must
+  #   be a subset of the user's accessible miq_product_features or :all if you wish to
+  #   share all the user's accessible features. Defaults to :all.
+  def initialize(args = {})
+    args = args.reverse_merge(:features => :all)
+    if args[:user] && args[:features] == :all
+      args[:features] = args[:user].miq_user_role.miq_product_features
+    end
+    super
+  end
+
+  def share
+    return false unless valid?
+
+    ActiveRecord::Base.transaction do
+      tenants.map do |tenant|
+        Share.create!(:user => user, :resource => resource, :tenant => tenant, :miq_product_features => features)
+      end
+    end
+  end
+
+  private
+
+  def rbac_visibility
+    return unless user && resource
+    unless Rbac::Filterer.search(:targets => [resource], :user => user)[0].include?(resource)
+      errors.add(:user, "is not authorized to share this resource")
+    end
```

Comment:
Lets avoid `search` if you don't need the second parameter.
Have introduced `filtered_object` to support rbac of a single object. (And also be a flag to let us know that we are filtering a single object and probably shouldn't.)

```ruby
unless Rbac::Filterer.filtered_object(resource, :user => user).exist?
  errors.add(:user, "is not authorized to share this resource")
end
```

`Rbac` is currently built to be a sql query of multiple objects, not called on each object one by one.
Can we share multiple objects in a go instead of verifying each individual object?

This is a form object to make data entry easier right?
I suppose we could have a form with multiple objects and verify them all in 1 go.
Also only looking up the user's privileges in one shot as well